### PR TITLE
Add support AppSync Function Resource and pipeline configuration for resolver

### LINKED ISSUE #8127: Add support for AppSync pipeline resolvers
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Trying out the new AppSync resolvers that were added in 2.4.0 (#2705) but I'm not seeing a way to configure a pipeline resolver with multiple functions in the current implementation of `appsync_resolver`. Seems like the AWS Provider would need a new `appsync_function` resource and a modification to `appsync_resolver` allowing users to pass a pipeline_config.

https://docs.aws.amazon.com/appsync/latest/devguide/pipeline-resolvers.html

### New or Affected Resource(s)

* appsync_function
* appsync_resolver

### Potential Terraform Configuration

```hcl
resource "aws_appsync_graphql_api" "test" {
  authentication_type = "API_KEY"
  name                = "tf-example"
  schema              = <<EOF
type Mutation {
    putPost(id: ID!, title: String!): Post
}

type Post {
    id: ID!
    title: String!
}

type Query {
    singlePost(id: ID!): Post
}

schema {
    query: Query
    mutation: Mutation
}
EOF
}

resource "aws_appsync_datasource" "test" {
  api_id      = "${aws_appsync_graphql_api.test.id}"
  name        = "tf-example"
  type        = "HTTP"

  http_config {
    endpoint = "http://example.com"
  }
}

resource "aws_appsync_function" "test" {
  api_id      = "${aws_appsync_graphql_api.test.id}"
  data_source = "${aws_appsync_datasource.test.name}"
  name        = "tf-example"
  request_mapping_template = <<EOF
{
    "version": "2018-05-29",
    "method": "GET",
    "resourcePath": "/",
    "params":{
        "headers": $utils.http.copyheaders($ctx.request.headers)
    }
}
EOF
  response_mapping_template = <<EOF
#if($ctx.result.statusCode == 200)
    $ctx.result.body
#else
    $utils.appendError($ctx.result.body, $ctx.result.statusCode)
#end
EOF
}

resource "aws_appsync_resolver" "test" {
  api_id           = "${aws_appsync_graphql_api.test.id}"
  field            = "singlePost"
  type             = "Query"
  data_source      = "${aws_appsync_datasource.test.name}"
  request_template = <<EOF
{
    "version": "2018-05-29",
    "method": "GET",
    "resourcePath": "/",
    "params":{
        "headers": $utils.http.copyheaders($ctx.request.headers)
    }
}
EOF
  response_template = <<EOF
#if($ctx.result.statusCode == 200)
    $ctx.result.body
#else
    $utils.appendError($ctx.result.body, $ctx.result.statusCode)
#end
EOF

  pipeline_config  = {
    functions = ["${aws_appsync_function.test.name}"]
  }
}
```

### References

* #2705

### PR DESCRIPTION
**Pull request by kterada0509**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #8127

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
- New Resource: aws_appsync_function (#8127)
- resource/aws_appsync_resolver: Add `pipeline_config` attribute (#8127)
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppsyncFunction_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAwsAppsyncFunction_ -timeout 120m
=== RUN   TestAccAwsAppsyncFunction_basic
=== PAUSE TestAccAwsAppsyncFunction_basic
=== RUN   TestAccAwsAppsyncFunction_description
=== PAUSE TestAccAwsAppsyncFunction_description
=== RUN   TestAccAwsAppsyncFunction_responseMappingTemplate
=== PAUSE TestAccAwsAppsyncFunction_responseMappingTemplate
=== RUN   TestAccAwsAppsyncFunction_disappears
=== PAUSE TestAccAwsAppsyncFunction_disappears
=== CONT  TestAccAwsAppsyncFunction_basic
=== CONT  TestAccAwsAppsyncFunction_disappears
=== CONT  TestAccAwsAppsyncFunction_responseMappingTemplate
=== CONT  TestAccAwsAppsyncFunction_description
--- PASS: TestAccAwsAppsyncFunction_disappears (73.59s)
--- PASS: TestAccAwsAppsyncFunction_description (95.39s)
--- PASS: TestAccAwsAppsyncFunction_basic (96.09s)
--- PASS: TestAccAwsAppsyncFunction_responseMappingTemplate (139.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	139.963s
```

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsAppsyncResolver_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAwsAppsyncResolver_ -timeout 120m
=== RUN   TestAccAwsAppsyncResolver_basic
=== PAUSE TestAccAwsAppsyncResolver_basic
=== RUN   TestAccAwsAppsyncResolver_disappears
=== PAUSE TestAccAwsAppsyncResolver_disappears
=== RUN   TestAccAwsAppsyncResolver_DataSource
=== PAUSE TestAccAwsAppsyncResolver_DataSource
=== RUN   TestAccAwsAppsyncResolver_RequestTemplate
=== PAUSE TestAccAwsAppsyncResolver_RequestTemplate
=== RUN   TestAccAwsAppsyncResolver_ResponseTemplate
=== PAUSE TestAccAwsAppsyncResolver_ResponseTemplate
=== RUN   TestAccAwsAppsyncResolver_multipleResolvers
=== PAUSE TestAccAwsAppsyncResolver_multipleResolvers
=== RUN   TestAccAwsAppsyncResolver_PipelineConfig
=== PAUSE TestAccAwsAppsyncResolver_PipelineConfig
=== CONT  TestAccAwsAppsyncResolver_basic
=== CONT  TestAccAwsAppsyncResolver_ResponseTemplate
=== CONT  TestAccAwsAppsyncResolver_PipelineConfig
=== CONT  TestAccAwsAppsyncResolver_multipleResolvers
=== CONT  TestAccAwsAppsyncResolver_RequestTemplate
=== CONT  TestAccAwsAppsyncResolver_disappears
=== CONT  TestAccAwsAppsyncResolver_DataSource
--- PASS: TestAccAwsAppsyncResolver_disappears (34.36s)
--- PASS: TestAccAwsAppsyncResolver_PipelineConfig (37.39s)
--- PASS: TestAccAwsAppsyncResolver_basic (40.98s)
--- PASS: TestAccAwsAppsyncResolver_RequestTemplate (52.62s)
--- PASS: TestAccAwsAppsyncResolver_ResponseTemplate (58.14s)
--- PASS: TestAccAwsAppsyncResolver_DataSource (59.19s)
--- PASS: TestAccAwsAppsyncResolver_multipleResolvers (62.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	62.168s
```

### CODE REVIEW

**Comment by ouadie-lahdioui** [comment]
+1

**Comment by witooh** [comment]
👍

**Comment by nqmt** [comment]
👍

**Comment by aholicz1208** [comment]
👍

**Comment by atthapon-k** [comment]
👍

**Comment by pichaya1450** [comment]
👍

**Comment by Amornthep** [comment]
👍

**Comment by laurentlouk** [comment]
Why this takes so long ? x)

**Comment by rimutaka** [comment]
I successfully used this PR in a custom build with no apparent issues.

File: aws/resource_aws_appsync_function_test.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+		input := &appsync.GetFunctionInput{
+			ApiId:      aws.String(apiID),
+			FunctionId: aws.String(functionID),
+		}
+
+		_, err = conn.GetFunction(input)
+		if err != nil {
+			if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+				return nil
```
Comment (bflad) [inline aws/resource_aws_appsync_function_test.go]: We should fix this in the future:

```suggestion
				continue
```

**Comment by bflad** [comment]
This has been released in [version 2.17.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2170-june-28-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!

**Pull request merged**